File: packages/sdk/java/packages/core/src/main/java/dev/edgebase/sdk/core/StorageBucket.java
// EdgeBase Java SDK — Storage bucket operations.
package dev.edgebase.sdk.core;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * Storage bucket — operations on files within a bucket.
 */
public class StorageBucket {
    private final HttpClient client;
    private final String name;

    private static String encodePathParam(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20");
    }

    StorageBucket(HttpClient client, String name) {
        this.client = client;
        this.name = name;
    }

    public String getName() {
        return name;
    }

    // ─── Upload ───

    @SuppressWarnings("unchecked")
    public FileInfo upload(String key, byte[] data) {
        return upload(key, data, null, null);
    }

    @SuppressWarnings("unchecked")
    public FileInfo upload(String key, byte[] data, String contentType) {
        return upload(key, data, contentType, null);
    }

    @SuppressWarnings("unchecked")
    public FileInfo upload(String key, byte[] data, String contentType,
            Map<String, String> customMetadata) {
        Map<String, String> extra = new LinkedHashMap<>();
        extra.put("key", key);
        if (contentType != null)
            extra.put("contentType", contentType);
        if (customMetadata != null) {
            for (Map.Entry<String, String> entry : customMetadata.entrySet()) {
                extra.put("meta_" + entry.getKey(), entry.getValue());
            }
        }

        String fileName = key.contains("/") ? key.substring(key.lastIndexOf('/') + 1) : key;
        Map<String, Object> json = (Map<String, Object>) client.uploadMultipart(
                "/storage/" + name + "/upload",
                fileName,
                data,
                contentType != null ? contentType : "application/octet-stream",
                extra);
        return FileInfo.fromJson(json);
    }

    /**
     * Upload string data with encoding.
     *
     * @param encoding One of: "raw", "base64", "base64url", "data_url"
     */
    public FileInfo uploadString(String key, String content) {
        return uploadString(key, content, "raw");
    }

    @SuppressWarnings("unchecked")
    public FileInfo uploadString(String key, String content, String encoding) {
        byte[] data;
        switch (encoding) {
            case "base64":
                data = Base64.getDecoder().decode(content);
                break;
            case "base64url":
                data = Base64.getUrlDecoder().decode(content);
                break;
            case "data_url":
                int commaIdx = content.indexOf(",");
                if (commaIdx >= 0) {
                    data = Base64.getDecoder().decode(content.substring(commaIdx + 1));
                } else {
                    data = content.getBytes(StandardCharsets.UTF_8);
                }
                break;
            default: // "raw"
                data = content.getBytes(StandardCharsets.UTF_8);
                break;
        }
        return upload(key, data);
    }

    // ─── Download ───

    public byte[] download(String key) {
        return client.downloadRaw("/storage/" + name + "/" +
                encodePathParam(key));
    }

    // ─── Delete ───

    public void delete(String key) {
        client.delete("/storage/" + name + "/" +
                encodePathParam(key));
    }

    // ─── List ───

    @SuppressWarnings("unchecked")
    public Map<String, Object> list() {
        return list(null, null, null);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> list(String prefix) {
        return list(prefix, null, null);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> list(String prefix, Integer limit, String cursor) {
        Map<String, String> params = new LinkedHashMap<>();
        if (prefix != null)
            params.put("prefix", prefix);
        if (limit != null)
            params.put("limit", limit.toString());
        if (cursor != null)
            params.put("cursor", cursor);
        return (Map<String, Object>) client.get("/storage/" + name,
                params.isEmpty() ? null : params);
    }

    // ─── URL ───

    public String getUrl(String key) {
        return client.getApiBaseUrl() + "/storage/" + name + "/" +
                encodePathParam(key);
    }

    // ─── Metadata ───

    @SuppressWarnings("unchecked")
    public FileInfo getMetadata(String key) {
        Map<String, Object> json = (Map<String, Object>) client.get(
                "/storage/" + name + "/" + encodePathParam(key) + "/metadata");
        return FileInfo.fromJson(json);
    }

    @SuppressWarnings("unchecked")
    public FileInfo updateMetadata(String key, Map<String, Object> metadata) {
        Map<String, Object> json = (Map<String, Object>) client.patch(
                "/storage/" + name + "/" + encodePathParam(key) + "/metadata",
                metadata);
        return FileInfo.fromJson(json);
    }

    // ─── Signed URLs ───

    @SuppressWarnings("unchecked")
    public SignedUrlResult createSignedUrl(String key) {
        return createSignedUrl(key, "1h");
    }

    @SuppressWarnings("unchecked")
    public SignedUrlResult createSignedUrl(String key, String expiresIn) {
        Map<String, Object> json = (Map<String, Object>) client.post(
                "/storage/" + name + "/signed-url",
                Map.of("key", key, "expiresIn", expiresIn));
        return new SignedUrlResult(
                (String) json.getOrDefault("url", ""),
                json.get("expiresIn") instanceof Number ? ((Number) json.get("expiresIn")).intValue() : 3600);
    }

    public List<SignedUrlResult> createSignedUrls(List<String> keys) {
        return createSignedUrls(keys, "1h");
    }

    @SuppressWarnings("unchecked")
    public List<SignedUrlResult> createSignedUrls(List<String> keys, String expiresIn) {
        Map<String, Object> json = (Map<String, Object>) client.post(
                "/storage/" + name + "/signed-urls",
                Map.of("keys", keys, "expiresIn", expiresIn));
        List<SignedUrlResult> results = new ArrayList<>();
        Object urls = json.get("urls");
        if (urls instanceof List<?>) {
            for (Object entry : (List<?>) urls) {
                if (!(entry instanceof Map<?, ?> map))
                    continue;
                Object url = map.get("url");
                Object rawExpires = map.get("expiresIn");
                results.add(new SignedUrlResult(
                        url instanceof String ? (String) url : "",
                        rawExpires instanceof Number ? ((Number) rawExpires).intValue() : 3600));
            }
        }
        return results;
    }

    @SuppressWarnings("unchecked")
    public SignedUrlResult createSignedUploadUrl(String key) {
        return createSignedUploadUrl(key, 3600);
    }

    @SuppressWarnings("unchecked")
    public SignedUrlResult createSignedUploadUrl(String key, int expiresIn) {
        Map<String, Object> json = (Map<String, Object>) client.post(
                "/storage/" + name + "/signed-upload-url",
                Map.of("key", key, "expiresIn", expiresIn + "s"));
        return new SignedUrlResult(
                (String) json.getOrDefault("url", ""),
                json.get("expiresIn") instanceof Number ? ((Number) json.get("expiresIn")).intValue() : expiresIn);
    }

    public boolean exists(String key) {
        return client.head("/storage/" + name + "/" + encodePathParam(key));
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> getUploadParts(String key, String uploadId) {
        Map<String, Object> json = (Map<String, Object>) client.get(
                "/storage/" + name + "/uploads/" + uploadId + "/parts",
                Map.of("key", key));
        Map<String, Object> normalized = new LinkedHashMap<>();
        normalized.put("uploadId", json.getOrDefault("uploadId", uploadId));
        normalized.put("key", json.getOrDefault("key", key));
        normalized.put("parts", json.getOrDefault("parts", Collections.emptyList()));
        return normalized;
    }

    // ─── Resumable Upload ───

    @SuppressWarnings("unchecked")
    public Map<String, Object> initiateResumableUpload(String key) {
        return initiateResumableUpload(key, null);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> initiateResumableUpload(String key, String contentType) {
        return (Map<String, Object>) client.post("/storage/" + name + "/multipart/create",
                Map.of("key", key, "contentType", contentType != null ? contentType : "application/octet-stream"));
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> resumeUpload(String key, String uploadId, byte[] chunk, long offset) {
        int partNumber = (int) (offset / (5L * 1024L * 1024L)) + 1;
        Map<String, Object> uploadedPart = (Map<String, Object>) client.postBytes(
                "/storage/" + name + "/multipart/upload-part",
                chunk,
                "application/octet-stream",
                Map.of(
                        "uploadId", uploadId,
                        "partNumber", String.valueOf(partNumber),
                        "key", key));

        String etag = uploadedPart.get("etag") instanceof String ? (String) uploadedPart.get("etag") : null;
        if (etag == null || etag.isEmpty()) {
            throw new EdgeBaseError(0, "Multipart upload missing etag");
        }

        return (Map<String, Object>) client.post(
                "/storage/" + name + "/multipart/complete",
                Map.of(
                        "uploadId", uploadId,
                        "key", key,
                        "parts", List.of(Map.of(
                                "partNumber", uploadedPart.get("partNumber") instanceof Number
                                        ? ((Number) uploadedPart.get("partNumber")).intValue()
                                        : partNumber,
                                "etag", etag))));
    }
}

File: packages/sdk/cpp/packages/core/src/storage_bucket.cpp
// EdgeBase C++ Core — StorageBucket implementation.
// JSON-based API calls delegate to GeneratedDbApi (api_core.h).
// Binary upload/multipart operations remain on HttpClient directly.
#include "edgebase/edgebase.h"
#include <nlohmann/json.hpp>

namespace client {

// ── StorageBucket
// ─────────────────────────────────────────────────────────────

StorageBucket::StorageBucket(std::shared_ptr<HttpClient> http,
                             std::shared_ptr<GeneratedDbApi> core,
                             std::string baseUrl, std::string name)
    : http_(std::move(http)), core_(std::move(core)),
      baseUrl_(std::move(baseUrl)), name_(std::move(name)) {}

std::string StorageBucket::getUrl(const std::string &key) const {
  return baseUrl_ + ApiPaths::download_file(name_, key);
}

Result StorageBucket::upload(const std::string &key,
                             const std::vector<uint8_t> &data,
                             const std::string &contentType) const {
  // Binary multipart upload — not available in GeneratedDbApi.
  return http_->uploadMultipart("/storage/" + name_ + "/upload", key, data,
                                contentType);
}

Result StorageBucket::uploadString(const std::string &key,
                                   const std::string &content,
                                   const std::string &encoding,
                                   const std::string &contentType) const {
  std::vector<uint8_t> data;
  auto ct = contentType;

  // Helper: base64 decode (standard)
  auto b64decode = [](const std::string &in) -> std::vector<uint8_t> {
    static const std::string chars =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    std::vector<uint8_t> out;
    int val = 0, bits = -8;
    for (unsigned char c : in) {
      if (c == '=')
        break;
      auto pos = chars.find(c);
      if (pos == std::string::npos)
        continue;
      val = (val << 6) + static_cast<int>(pos);
      bits += 6;
      if (bits >= 0) {
        out.push_back(static_cast<uint8_t>((val >> bits) & 0xFF));
        bits -= 8;
      }
    }
    return out;
  };

  if (encoding == "base64") {
    data = b64decode(content);
  } else if (encoding == "base64url") {
    // Convert base64url to standard base64
    std::string b64 = content;
    for (auto &c : b64) {
      if (c == '-')
        c = '+';
      else if (c == '_')
        c = '/';
    }
    while (b64.size() % 4 != 0)
      b64 += '=';
    data = b64decode(b64);
  } else if (encoding == "data_url") {
    auto commaPos = content.find(',');
    if (commaPos != std::string::npos) {
      // Extract MIME type from header: data:mime/type;base64
      auto header = content.substr(0, commaPos);
      if (header.rfind("data:", 0) == 0) {
        auto semiPos = header.find(';');
        if (semiPos != std::string::npos && semiPos > 5) {
          ct = header.substr(5, semiPos - 5);
        }
      }
      data = b64decode(content.substr(commaPos + 1));
    } else {
      data.assign(content.begin(), content.end());
    }
  } else { // "raw"
    data.assign(content.begin(), content.end());
  }

  return upload(key, data, ct);
}

Result StorageBucket::download(const std::string &key) const {
  return core_->download_file(name_, key);
}

Result StorageBucket::del(const std::string &key) const {
  return core_->delete_file(name_, key);
}

// GET /api/storage/{bucket}?prefix=&limit=N&offset=N
Result StorageBucket::list(const std::string &prefix, int limit,
                           int offset) const {
  // GeneratedDbApi::list_files() doesn't support query params (prefix/limit/offset),
  // so we fall back to HttpClient for parameterized list requests.
  if (prefix.empty() && limit == 100 && offset == 0) {
    return core_->list_files(name_);
  }
  std::string path = "/storage/" + name_;
  std::string qs;
  if (!prefix.empty())
    qs += "prefix=" + prefix;
  if (limit > 0) {
    if (!qs.empty())
      qs += "&";
    qs += "limit=" + std::to_string(limit);
  }
  if (offset > 0) {
    if (!qs.empty())
      qs += "&";
    qs += "offset=" + std::to_string(offset);
  }
  if (!qs.empty())
    path += "?" + qs;
  return http_->get(path);
}

Result StorageBucket::getMetadata(const std::string &key) const {
  return core_->get_file_metadata(name_, key);
}

Result StorageBucket::updateMetadata(const std::string &key,
                                     const std::string &jsonBody) const {
  return core_->update_file_metadata(name_, key, jsonBody);
}

Result StorageBucket::createSignedUrl(const std::string &key,
                                      const std::string &expiresIn) const {
  nlohmann::json body = {{"key", key}, {"expiresIn", expiresIn}};
  return core_->create_signed_download_url(name_, body.dump());
}

Result StorageBucket::createSignedUrls(const std::vector<std::string> &keys,
                                       const std::string &expiresIn) const {
  nlohmann::json body = {{"keys", keys}, {"expiresIn", expiresIn}};
  return core_->create_signed_download_urls(name_, body.dump());
}

Result
StorageBucket::createSignedUploadUrl(const std::string &key,
                                     const std::string &expiresIn) const {
  nlohmann::json body = {{"key", key}, {"expiresIn", expiresIn}};
  return core_->create_signed_upload_url(name_, body.dump());
}

bool StorageBucket::exists(const std::string &key) const {
  return core_->check_file_exists(name_, key);
}

Result StorageBucket::getUploadParts(const std::string &key,
                                     const std::string &uploadId) const {
  auto result = http_->get(
      "/storage/" + name_ + "/uploads/" + uploadId + "/parts",
      {{"key", key}});
  if (!result.ok)
    return result;
  auto body = nlohmann::json::parse(result.body, nullptr, false);
  if (body.is_discarded() || !body.is_object()) {
    return {false, result.statusCode, result.body, "Invalid getUploadParts response"};
  }
  if (!body.contains("uploadId"))
    body["uploadId"] = uploadId;
  if (!body.contains("key"))
    body["key"] = key;
  if (!body.contains("parts"))
    body["parts"] = nlohmann::json::array();
  result.body = body.dump();
  return result;
}

Result
StorageBucket::initiateResumableUpload(const std::string &key,
                                       const std::string &contentType) const {
  nlohmann::json body = {{"key", key}};
  if (!contentType.empty())
    body["contentType"] = contentType;
  return core_->create_multipart_upload(name_, body.dump());
}

Result StorageBucket::resumeUpload(const std::string &key,
                                   const std::string &uploadId,
                                   const std::vector<uint8_t> &chunk,
                                   int offset, bool isLastChunk) const {
  const int partNumber = offset + 1;
  auto uploaded = http_->post_bytes_with_query(
      "/storage/" + name_ + "/multipart/upload-part", chunk,
      "application/octet-stream",
      {{"uploadId", uploadId},
       {"partNumber", std::to_string(partNumber)},
       {"key", key}});
  if (!uploaded.ok || !isLastChunk)
    return uploaded;

  const auto uploadedBody =
      nlohmann::json::parse(uploaded.body, nullptr, false);
  if (uploadedBody.is_discarded() || !uploadedBody.is_object()) {
    return {false, uploaded.statusCode, uploaded.body,
            "Invalid multipart upload-part response"};
  }

  nlohmann::json completeBody = {
      {"uploadId", uploadId},
      {"key", key},
      {"parts", nlohmann::json::array(
                    {{{"partNumber", uploadedBody.value("partNumber", partNumber)},
                      {"etag", uploadedBody.value("etag", "")}}})},
  };
  return core_->complete_multipart_upload(name_, completeBody.dump());
}

// ── StorageClient
// ─────────────────────────────────────────────────────────────

StorageClient::StorageClient(std::shared_ptr<HttpClient> http,
                             std::shared_ptr<GeneratedDbApi> core,
                             std::string baseUrl)
    : http_(std::move(http)), core_(std::move(core)),
      baseUrl_(std::move(baseUrl)) {}

StorageBucket StorageClient::bucket(const std::string &name) const {
  return StorageBucket(http_, core_, baseUrl_, name);
}

} // namespace client
